internal/cmd/io_binary.go

```diff
@@ -0,0 +1,254 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/url"
+	"os/exec"
+	"sync"
+	"time"
+
+	"github.com/Microsoft/go-winio"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/pkg/errors"
+)
+
+const (
+	pipeRootBinary       = `\\.\pipe`
+	binaryCmdWaitTimeout = 5 * time.Second
+)
+
+func newBinaryCmd(ctx context.Context, uri *url.URL, id string, ns string) *exec.Cmd {
+	var args []string
+	for k, vs := range uri.Query() {
+		args = append(args, k)
+		if len(vs) > 0 {
+			args = append(args, vs[0])
+		}
+	}
+
+	execPath := uri.Path
+
+	cmd := exec.CommandContext(ctx, execPath, args...)
+	cmd.Env = append(cmd.Env,
+		"CONTAINER_ID="+id,
+		"CONTAINER_NAMESPACE="+ns,
+	)
+
+	return cmd
+}
+
+type binaryIO struct {
+	cmd *exec.Cmd
+
+	binaryCloser sync.Once
+
+	stdin, stdout, stderr string
+
+	sout, serr, wait io.ReadWriteCloser
+	soutCloser       sync.Once
+}
+
+// NewBinaryIO starts a binary logger
+func NewBinaryIO(ctx context.Context, id string, uri *url.URL) (_ UpstreamIO, err error) {
+	ns, err := namespaces.NamespaceRequired(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		sout, serr, w io.ReadWriteCloser
+	)
+
+	bio := &binaryIO{}
+
+	stdoutPipe := fmt.Sprintf(`%s\binary-%s-stdout`, pipeRootBinary, id)
+	sout, err = openNPipe(stdoutPipe)
+	if err != nil {
+		return nil, err
+	}
+	bio.stdout = stdoutPipe
+	bio.sout = sout
+
+	stderrPipe := fmt.Sprintf(`%s\binary-%s-stderr`, pipeRootBinary, id)
+	serr, err = openNPipe(stderrPipe)
+	if err != nil {
+		return nil, err
+	}
+	bio.stderr = stderrPipe
+	bio.serr = serr
+
+	waitPipe := fmt.Sprintf(`%s\binary-%s-wait`, pipeRootBinary, id)
+	w, err = openNPipe(waitPipe)
+	if err != nil {
+		return nil, err
+	}
+	bio.wait = w
+
+	cmd := newBinaryCmd(ctx, uri, id, ns)
+	cmd.Env = append(cmd.Env,
+		"CONTAINER_STDOUT="+stdoutPipe,
+		"CONTAINER_STDERR="+stderrPipe,
+		"CONTAINER_WAIT="+waitPipe,
+	)
+
+	bio.cmd = cmd
+
+	started := make(chan bool, 1)
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+
+	go func() {
+		defer w.Close()
+		t := 0
+		for {
+			b := make([]byte, 1)
+			v, err := w.Read(b)
+			if (v == 0) || (err != nil && err != io.EOF) {
+				time.Sleep(1 * time.Second)
+				t++
+				if t > 10 {
+					break
+				}
+			} else {
+				started <- true
+				return
+			}
+		}
```

Comment:
If the logic here is essentially 

1. Try and read a byte 10 times
2. On failure sleep 1 second and increment a counter until we eventually break

we could just make the for loop loop 10 times explicitly and add the timer at the end. This was a bit hard to follow at first glance so it would help readability a bit. 

```go
for i:=0; i < 10; i++ {
	b := make([]byte, 1)
	v, err := w.Read(b)
	if v != 0 && err == nil {
             started <- true 
             return
        }
        time.Sleep(time.Second*1)
}	
started <- false	
```

Reply:
yeah, thanks